libsolidity/analysis/PostTypeContractLevelChecker.cpp

```diff
@@ -134,5 +136,20 @@ void PostTypeContractLevelChecker::checkStorageLayoutSpecifier(StorageLayoutSpec
 	}
 
 	solAssert(baseSlotExpressionType->isImplicitlyConvertibleTo(*TypeProvider::uint256()));
-	_storageLayoutSpecifier.annotation().baseSlot = u256(rationalType->value().numerator());
+	bigint baseSlot = u256(rationalType->value().numerator());
+	storageLayoutSpecifier->annotation().baseSlot = u256(baseSlot);
+
+	bigint size = 0;
+	for (ContractDefinition const* contract: _contract.annotation().linearizedBaseContracts)
+		for (VariableDeclaration const* variable: contract->stateVariables())
+			if (!(variable->isConstant() || variable->immutable()))
+				size += variable->annotation().type->storageSizeUpperBound();
```

Comment:
This calculation is also in `ContractLevelChecker::checkStorageSize()`. If we need it more than once, we should make it a helper.

`ContractDefinition::storageSizeUpperBound()` could be a good place. Though then it may be too easy to confuse with `ContractType::storageSizeUpperBound()` (which returns 1) so perhaps something like `contractStorageSizeUpperBound()` in `ASTUtils.h` is a better idea.

Reply:
Oh, I think both of those checks are also buggy. This code does not distinguish between transient and non-transient variables so it probably adds up their sizes, which is wrong.

For `ContractLevelChecker::checkStorageSize()` that needs a changelog entry and should be fixed in a separate PR.

In your case it's actually more severe, because it can be easily triggered by an unrelated transient variable even when things are reasonably sized. It's enough that the layout places things close to the end of storage. Please make sure you have tests covering this.

Reply:
#15899

Reply:
You can now use the helper you added in #15901.